pkg/controller/sparkapplication/sparkapp_metrics.go

```diff
@@ -0,0 +1,235 @@
+/*
+Copyright 2018 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sparkapplication
+
+import (
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"k8s.io/spark-on-k8s-operator/pkg/apis/sparkoperator.k8s.io/v1alpha1"
+	"k8s.io/spark-on-k8s-operator/pkg/util"
+)
+
+type sparkAppMetrics struct {
+	labels []string
+	prefix string
+
+	sparkAppSubmitCount  *prometheus.CounterVec
+	sparkAppSuccessCount *prometheus.CounterVec
+	sparkAppFailureCount *prometheus.CounterVec
+	sparkAppRunningCount *util.PositiveGauge
+
+	sparkAppSuccessExecutionTime *prometheus.SummaryVec
+	sparkAppFailureExecutionTime *prometheus.SummaryVec
+
+	sparkAppExecutorRunningCount *util.PositiveGauge
+	sparkAppExecutorFailureCount *prometheus.CounterVec
+	sparkAppExecutorSuccessCount *prometheus.CounterVec
+}
+
+func newSparkAppMetrics(prefix string, labels []string) *sparkAppMetrics {
+	for i, label := range labels {
```

Comment:
To be safe, let's make and use a copy of `labels` here. The in-place changes to `labels` here are visible outside the caller of this function.

Reply:
done